codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/EndpointDelegator.kt

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package software.amazon.smithy.kotlin.codegen.rendering.endpoints
+
+import software.amazon.smithy.kotlin.codegen.core.useFileWriter
+import software.amazon.smithy.kotlin.codegen.rendering.protocol.ProtocolGenerator
+import software.amazon.smithy.rulesengine.language.EndpointRuleSet
+import software.amazon.smithy.rulesengine.traits.EndpointTestCase
+
+/**
+ * Responsible for providing various endpoint related generation components
+ */
+interface EndpointDelegator {
+    companion object {
+        /**
+         * Get the default endpoint delegator
+         */
+        val Default: EndpointDelegator = object : EndpointDelegator {}
+    }
+    /**
+     * Generate a default implementation for the modeled endpoint provider.
+     * Will only be invoked when a model's service shape has the necessary endpoint rule set trait, however, the base
+     * interface and parameter type will always be generated (the expectation being that the caller supplies their own
+     * at runtime).
+     */
+    fun generateEndpointProvider(ctx: ProtocolGenerator.GenerationContext, rules: EndpointRuleSet?) {
+        val paramsSymbol = EndpointParametersGenerator.getSymbol(ctx.settings)
+        val providerSymbol = EndpointProviderGenerator.getSymbol(ctx.settings)
+        val defaultProviderSymbol = DefaultEndpointProviderGenerator.getSymbol(ctx.settings)
+
+        ctx.delegator.useFileWriter(providerSymbol) {
+            EndpointProviderGenerator(it, paramsSymbol).render()
+        }
+
+        if (rules != null) {
+            ctx.delegator.useFileWriter(defaultProviderSymbol) {
+                DefaultEndpointProviderGenerator(it, rules, providerSymbol, paramsSymbol).render()
+            }
+        }
+    }
+
+    /**
+     * Generate the input parameter type for the generated endpoint provider implementation.
+     */
+    fun generateEndpointParameters(ctx: ProtocolGenerator.GenerationContext, rules: EndpointRuleSet?) {
+        val paramsSymbol = EndpointParametersGenerator.getSymbol(ctx.settings)
+        ctx.delegator.useFileWriter(paramsSymbol) {
+            EndpointParametersGenerator(it, rules).render()
+        }
+    }
+
+    /**
+     * Generate an implementation and supporting code for the modeled endpoint provider.
```

Comment:
This just generates the test cases right? Not the implementation and support code for the endpoint provider?